OSDOCS-10036:Exporting Network Observability metrics to OpenTelemetry (#80555)
openshift/openshift-docs: modules/network-observability-enriched-flows.adoc

[id="network-observability-enriched-flows_{context}"]
= Export enriched network flow data

> You can send network flows to Kafka, IPFIX, the Red Hat build of OpenTelemetry, or all three at the same time. For Kafka or IPFIX, any processor or storage that supports those inputs, such as Splunk, Elasticsearch, or Fluentd, can consume the enriched network flow data.

Reviewer: Actually, OpenTelemetry is a framework that can be present in any app such as Jaeger and Prometheus.
https://opentelemetry.io/docs/what-is-opentelemetry/

Netobserv allows to export either flows and / or metrics to a compatible OpenTelemetry endpoint; even if it's not a Red Hat product

> You can send network flows to Kafka, IPFIX, the Red Hat build of OpenTelemetry, or all three at the same time. For Kafka or IPFIX, any processor or storage that supports those inputs, such as Splunk, Elasticsearch, or Fluentd, can consume the enriched network flow data. For OpenTelemetry, network flow data and metrics can be exported to a compatible OpenTelemetry endpoint, such as Red Hat build of OpenTelemetry, Jaeger, or Prometheus. 

Reviewer 2: Per the [SSG](https://redhat-documentation.github.io/supplementary-style-guide/#non-breaking-spaces), there should be a non-breaking space between Red and Hat: 
```suggestion
You can send network flows to Kafka, IPFIX, the Red{nbsp}Hat build of OpenTelemetry, or all three at the same time. For Kafka or IPFIX, any processor or storage that supports those inputs, such as Splunk, Elasticsearch, or Fluentd, can consume the enriched network flow data. For OpenTelemetry, network flow data and metrics can be exported to a compatible OpenTelemetry endpoint, such as Red{nbsp}Hat build of OpenTelemetry, Jaeger, or Prometheus. 
```

Author: Ahh this is nifty. I don't usually use "Red Hat" in NetObserv docs, so I didn't know this! Thanks :)

.Prerequisites
* Your Kafka, IPFIX, or OpenTelemetry collector endpoint(s) are available from Network Observability `flowlogs-pipeline` pods.

> * If you are exporting to OpenTelemetry, you must install the Red Hat build of OpenTelemetry Operator.

Reviewer: ```suggestion
```

Author: @jpinsonneau should this prerequisite instead be, 
> If you are exporting to OpenTelemetry, you must have a compatible OpenTelemetry endpoint available, such as Red Hat build of OpenTelemetry, Jaeger, or Prometheus.

Or no prerequisite at all?

Reviewer: ```
Your Kafka, IPFIX, or OpenTelemetry collector endpoint(s) are available from Network Observability `flowlogs-pipeline` pods.
```
:arrow_up:  That one is still valid

> * Your Kafka, IPFIX, or OpenTelemetry collector endpoint(s) are available from Network Observability `flowlogs-pipeline` pods.

Reviewer 2: Per the [ISG](https://www.ibm.com/docs/en/ibm-style?topic=punctuation-parentheses#plurals) we should avoid using (s) to indicate that something is potentially plural. In this case, I feel like making it plural makes sense:
```suggestion
* Your Kafka, IPFIX, or OpenTelemetry collector endpoints are available from Network Observability `flowlogs-pipeline` pods.
```

But totally up to you if you think another option is better

          enable: true
        metrics:                        <7>
          enable: true
          prefix: netobserv
          pushTimeInterval: 20s         <8>
          expiryTime: 2m
   #    fieldsMapping:                  <9>
   #      input: SrcAddr
   #      output: source.address
----
<1> You can export flows to IPFIX, OpenTelemetry, and Kafka individually or concurrently.
<2> The Network Observability Operator exports all flows to the configured Kafka topic.
<3> You can encrypt all communications to and from Kafka with SSL/TLS or mTLS. When enabled, the Kafka CA certificate must be available as a ConfigMap or a Secret, both in the namespace where the `flowlogs-pipeline` processor component is deployed (default: netobserv). It must be referenced with `spec.exporters.tls.caCert`. When using mTLS, client secrets must be available in these namespaces as well (they can be generated for instance using the AMQ Streams User Operator) and referenced with `spec.exporters.tls.userCert`. 
<4> You have the option to specify transport. The default value is `tcp` but you can also specify `udp`.
<5> The protocol of OpenTelemetry connection. The available options are `http` and `grpc`.

> <6> OpenTelemetry configuration for exporting logs, which are the same as the logs created by Loki. 
> <7> OpenTelemetry configuration for exporting logs, which are the same as the metrics created by Prometheus. These configurations are specified in the `spec.processor.metrics.includeList` parameter of the `FlowCollector` custom resource, along with any custom metrics you defined using the `FlowMetrics` custom resource.

Reviewer 3: by => for, I think (we create the logs/metrics, then store them in loki/prometheus)
also a copy-paste typo: logs => metrics on item 7

```suggestion
<6> OpenTelemetry configuration for exporting logs, which are the same as the logs created for Loki. 
<7> OpenTelemetry configuration for exporting metrics, which are the same as the metrics created for Prometheus. These configurations are specified in the `spec.processor.metrics.includeList` parameter of the `FlowCollector` custom resource, along with any custom metrics you defined using the `FlowMetrics` custom resource.
```

Author: Thank you! Good catch.

        targetHost: "ipfix-collector.ipfix.svc.cluster.local"
        targetPort: 4739
        transport: tcp or udp           <4>
 -  type: OpenTelemetry                 <3>
      openTelemetry:
        targetHost: my-otelcol-collector-headless.otlp.svc
        targetPort: 4317
        type: grpc                      <5>
        logs:                           <6>
          enable: true
        metrics:                        <7>
          enable: true
          prefix: netobserv
          pushTimeInterval: 20s         <8>
          expiryTime: 2m
        fieldsMapping:                  <9>
          input: source.address

>           output: SrcAddr

Reviewer: Since this is optionnal, should we put it in the example ?

The risk here is someone copy / pasting the example without reading the note

Author: I decided to add the OTEL in with what we've already got for Exporting enriches network flow data. I experimented with breaking them up, but I felt like ultimately adding it in here isn't so bad, especially since we'll add that column to the existing flow formats table and that info will just be a link. 

Maybe here, I can # out the fieldsMapping so if it is copied, its at least commented?

Reviewer: yeah that will work ! Thanks Sara

>    #    fieldsMapping:                  <9>
>    #      input: source.address
>    #      output: SrcAddr

Reviewer: this is actually the opposite !
```suggestion
   #    fieldsMapping:                  <9>
   #      input: SrcAddr
   #      output: source.address
```

Author: Oh! Idk how I got these mixed 🙃

>    #      output: source.address

Reviewer 2: Just checking, is this meant to be commented out in the YAML? Because it's showing up in the preview, just as commented out parameters

Reviewer: Most of the users **should not** apply the fieldsMapping. This is just an example of remapping for the few users concerned.

That's why we decided to comment it in the example. It will avoid support in case users copy/paste the example without fully reading the context.

Author: Sorry, I'm delayed in responding to peer review feedback but I just wanted to say yep what Julien says. And the callout indicates its optional. Its mostly just for awareness of the feature, but its not necessarily something a user might need or want to do.

----
<1> The Network Observability Operator exports all flows to the configured Kafka topic.
<2> You can encrypt all communications to and from Kafka with SSL/TLS or mTLS. When enabled, the Kafka CA certificate must be available as a ConfigMap or a Secret, both in the namespace where the `flowlogs-pipeline` processor component is deployed (default: netobserv). It must be referenced with `spec.exporters.tls.caCert`. When using mTLS, client secrets must be available in these namespaces as well (they can be generated for instance using the AMQ Streams User Operator) and referenced with `spec.exporters.tls.userCert`.

> <3> You can export flows to IPFIX, OpenTelemetry, and Kafka individually or concurrently. 

Reviewer 2: I guess this already existed this way before this PR, but it seems odd that the first item in the YAML with a callout has a callout of 3. Is it possible to bump this up to 1, so that there's numerical order in the YAML (at least whenever a unique labelled item appears in the file)?

Author: That's a good idea. I think it ended up this way when I added IPFIX to the mix, because previously this was just a topic about Kafka exporting. Its since become IPFIX, Kafka, and now OTEL. So yeah I think thats a good point to  rearrange these callouts to make them more numerically ordered.

<4> You have the option to specify transport. The default value is `tcp` but you can also specify `udp`.
<5> The protocol of OpenTelemetry connection. The available options are `http` and `grpc`.
<6> OpenTelemetry configuration for exporting logs, which are the same as the logs created by Loki. 

> <7> OpenTelemetry configuration for exporting logs, which are the same as the metrics created by Prometheus, which are specified in the `spec.processor.metrics.includeList` of the `FlowCollector` and any custom metrics you defined using the `FlowMetrics` custom resource.

Reviewer 2: I think this is a bit of a run-on sentence with the two "which are" phrases right next to each other. I would suggest breaking this up into two sentences:
```suggestion
<7> OpenTelemetry configuration for exporting logs, which are the same as the metrics created by Prometheus. These [metrics/configurations/whatever] are specified in the `spec.processor.metrics.includeList` parameter of the `FlowCollector` and any custom metrics you defined using the `FlowMetrics` custom resource.
```

Author: Good suggestion.

<8> The time interval that metrics are sent to the OpenTelemetry collector.

> <9> Network Observability network flows formats get automatically renamed to an OpenTelemetry compliant format. The `fieldsMapping` specification gives you the ability to rename these OpenTelemetry formats, for example, back to the original Network Observability format. In the YAML sample, `source.address` is the OpenTelemetry format and it is being renamed to the original Network Observability format, `SrcAddr` You can see both Network Observability and OpenTelemetry formats in the "Network flows format table".

Author: The Network flows format reference will be updated to with a column for the OTEL formats alongside the NetObserv formats.

Author: @jpinsonneau @jotak we just need to make sure that these OTEL formats get incorporated when we regenerate the Flows format reference and API references

Author: Ty :)

> <9> Network Observability network flows formats get automatically renamed to an OpenTelemetry compliant format. The `fieldsMapping` specification gives you the ability to rename these OpenTelemetry formats, for example, back to the original Network Observability format. In the YAML sample, `source.address` is the OpenTelemetry format and it is being renamed to the original Network Observability format, `SrcAddr` You can see both Network Observability and OpenTelemetry formats in the "Network flows format reference".

Reviewer 4: I don't think I have tried fieldsMapping, @jpinsonneau have you?

Reviewer: I have implemented it so yes I did :sunglasses: 
https://github.com/netobserv/network-observability-operator/pull/671#discussion_r1661190670

Reviewer: ```suggestion
<9> Network Observability network flows formats get automatically renamed to an OpenTelemetry compliant format. The `fieldsMapping` specification gives you the ability to customize the OpenTelemetry format output from the Network Observability format. For example in the YAML sample, `SrcAddr` is the Network Observability input field and it is being renamed to `source.address` in OpenTelemetry output. You can see both Network Observability and OpenTelemetry formats in the "Network flows format reference".
```

> . After configuration, network flows data can be sent to an available output in a JSON format. For more information, see "Network flows format reference".

Reviewer 2: This last step doesn't really instruct users to do anything, I think this can just be a regular sentence at the end of the procedure:
```suggestion

After configuration, network flows data can be sent to an available output in a JSON format. For more information, see "Network flows format reference".
```

Author: Agree, and removed this from the steps.
